pytorch_lightning/strategies/dp.py

```diff
@@ -137,18 +137,15 @@ def predict_step(self, *args, **kwargs) -> STEP_OUTPUT:
             return self.model(*args, **kwargs)
 
     def training_step_end(self, output):
-        if not is_overridden("training_step_end", self.lightning_module):
-            return self.reduce(output)
-        return output
+        if is_overridden("training_step_end", self.lightning_module):
+            return output
 
-    def validation_step_end(self, output):
```

Comment:
Why not reduce the evaluation outputs?

Reply:
the PR updates the logic to reduce only the loss.
reduction for loss is required to make sure `loss.backward` works perfectly for dp.
during evaluation we don't need a loss to compute gradients thus we don't reduce anything.

Reply:
one other way to do this would be I think to change `loss.backward()` to `loss.backward(torch.ones_like(loss))` (which should be mathematically equal) and not to reduce anything.

Reply:
TIL.

we can do this.. but just wondering
is this a common practice?
since we expose backward hook.

Reply:
I am not sure. But this is the way to handle non-scalar loss-values. We could also add some logic beforehand to check if `gradients` is passed in kwargs and add it there if not and if the loss is not scalar.

Reply:
Or let's keep it like this. Maybe not super convinient for every user to do it this way, since we ourselves have doubts of it being mathematically equivalent. If someone wants do configure this differently then can still choose to override `training_step_end`.

Reply:
Yes it is equivalent and as @justusschock said the key is the chain rule. 

Detailed:


<details>
  <summary>Details</summary>
  
Let x = f(w) be the unreduced output of a neural net f with weights w, where x = [x0, x1] has size 2 for this this illustration. Let L = x0 + x1 be the reduced loss. 

The derivative of L w.r.t x0 is 

dL / dx0 = 1

The derivative of L w.r.t. x1 is 

dL / dx1 = 1

So together: dL / dx = [1, 1]

The chain rule says:
dL / dw = (dL / dx) * (dx / dw)

With `x.backward([1, 1])` we tell pytorch that we have computed the value of dL/dx ourselves, i.e. [1, 1], and we would like pytorch to compute the rest, i.e., dx/dw so that finally we obtain the derivative dL/dw. 

So, L.backward() is the same as x.backward([1, 1]).
In DP, x0 gets computed by GPU 0, and x1 gets computed by GPU 1, but that doesn't change the math. 

</details>

I don't know if for sure we should do this but if it leads to a simplification, I think it would be fine to explore it (perhaps after this PR).

Reply:
IMO, I would go for Justus's proposal. I believe it is cleaner to leave the tensors exactly as they are and let the users deal with them.

Reply:
the reduced loss in this case will be: L = (x0 + x1) / 2, since batches are split across devices.

do we have to divide the loss tensor [x1, x2] by `num_devices` in such a case then?